Rakeify rake tasks

drbrain opened the pull request:
This PR makes the rake tasks in mruby-cli use more rake features than they presently do.

Most notable is use of [import](http://docs.seattlerb.org/rake/doc/rakefile_rdoc.html#label-Importing+Dependencies) to load the Rakefile from mruby after the appropriate setup has been finished (same for the package tasks).

Ensuring certain tasks are run inside Docker now uses dependencies instead of rewriting the tasks.  (Rewriting the tasks is bad because it converts FileTasks into Tasks and changes their behavior.)

Also, FileUtils references are removed as Rake already included that for us.

The result is as Jim Weirich intended.

drbrain committed:
Rake includes FileUtils at top-level for us

Also, remove () to be consistent across uses of FileUtils methods

drbrain committed:
Use cd from Rake instead of Dir.chdir

drbrain committed:
Extract expand_and_set method

This makes it more clear what the commend means

drbrain committed:
Use ENV.fetch with default value over ||

drbrain committed:
Use mruby_root for directory task

I think that previously the Rakefile would always create mruby_root as
"mruby" regardless of what was in the MRUBY_ROOT environment variable.
Now the correct directory will be created.

A `mruby` task is retained for compatibility.

drbrain committed:
Extract package tasks to tasks/package.rake

drbrain committed:
Make task to set APP_VERSION

This moves us a step closer to not having `cd mruby_root` polluting the
package tasks and allows us to make loading mruby use `import` as Jim
Weirich intended.

drbrain committed:
Use dependencies for checking for Docker

If we rewrite the task Bad Things happen (like a Rake::FileTask turning
into a Rake::Task).  Instead we can use a dependency because that runs
before anything the task does that would require Docker.

PS: rake-hooks is a pointless library.  Rake already has a way to run
things before a task, a prerequisite.  Rake already has a way to run
things after a task, make the after-task have the before-task a
prerequisite.

With some organization of tasks the before/after isn't an issue, it just
falls out with how you hook up your dependencies.

drbrain committed:
Use #import (as Jim Weirich intended)

Rake has the #import method to safely import Rakefiles into the current
rake application.  Using load or require may cause rakefiles to appear
in the wrong place or require cumbersome setup.

In this case the cumbersome setup was the `cd "mruby"` that required
adjustment for the correct path in the release and package tasks.

drbrain committed:
Make directory inside (or via) tasks

Otherwise we're making directories when they're not needed.

drbrain committed:
Run tests from mruby_root

Missed this one

drbrain commented:
I get the same test failure if I build an app with the latest release of mruby-cli and do nothing but run `docker-compose run test`:

```
RuntimeError: File TEST SETUP => can't make a socket bi (mrbgems: mruby-io)
Skip: File.expand_path (with ENV)  
RuntimeError: FileTest TEST SETUP => can't make a socket bi (mrbgems: mruby-io)
Fail: FileTest.socket? (mrbgems: mruby-io)
 - Assertion[1] Failed: Expected false to be true
    Expected: true
      Actual: false
RuntimeError: IO TEST SETUP => can't make a socket bi (mrbgems: mruby-io)
MTest 1) Skipped:
test_assert_match(Test4MTest) /home/mruby/code/mruby/build/mrbgems/mruby-mtest/mrblib/mtest_unit.rb:183: assert_match is not defined, because Regexp is not impl. (MTest::Skip)
Total: 687
   OK: 683
   KO: 1
Crash: 3
 Time: 0.13 seconds
rake aborted!
```

drbrain committed:
Move test tasks to test.rake to fix dependencies

The test task depends upon clearing MRuby's test task but we can't do
that until after MRuby's Rakefile loads.

Moving the test tasks to a separate file we import allows us to clear
MRuby's test actions at the proper time.

drbrain commented:
Figured it out, need to clear the `test` task of MRuby's actions after importing MRuby's Rakefile.

zzak commented:
The patch looks good to me, but we probably need to test this to make sure it doesn't break release or anything (I'm not sure those are thoroughly tested).

drbrain commented:
Seems fine:

```
$ docker-compose run release

Build summary:

================================================
      Config Name: host
 Output Directory: mruby/build/host
         Binaries: mrbc, mrbtest
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
             mruby-compiler - mruby compiler library
             mruby-bin-mrbc - mruby compiler executable
             mruby-test - mruby test
================================================

================================================
      Config Name: x86_64-pc-linux-gnu
 Output Directory: mruby/build/x86_64-pc-linux-gnu
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: i686-pc-linux-gnu
 Output Directory: mruby/build/i686-pc-linux-gnu
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: x86_64-apple-darwin14
 Output Directory: mruby/build/x86_64-apple-darwin14
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: i386-apple-darwin14
 Output Directory: mruby/build/i386-apple-darwin14
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: x86_64-w64-mingw32
 Output Directory: mruby/build/x86_64-w64-mingw32
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: i686-w64-mingw32
 Output Directory: mruby/build/i686-w64-mingw32
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

Writing /home/mruby/code/releases/v0.0.4/mruby-cli-0.0.4-x86_64-pc-linux-gnu.tgz
Writing /home/mruby/code/releases/v0.0.4/mruby-cli-0.0.4-i686-pc-linux-gnu.tgz
Writing /home/mruby/code/releases/v0.0.4/mruby-cli-0.0.4-x86_64-apple-darwin14.tgz
Writing /home/mruby/code/releases/v0.0.4/mruby-cli-0.0.4-i386-apple-darwin14.tgz
Writing /home/mruby/code/releases/v0.0.4/mruby-cli-0.0.4-x86_64-w64-mingw32.tgz
Writing /home/mruby/code/releases/v0.0.4/mruby-cli-0.0.4-i686-w64-mingw32.tgz
Writing /home/mruby/code/releases/v0.0.4/mruby-cli-0.0.4.tgz
```

PS: `package` seems broken on master:

```
$ docker-compose run package
[…]
Build summary:

================================================
      Config Name: host
 Output Directory: build/host
         Binaries: mrbc, mrbtest
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
             mruby-compiler - mruby compiler library
             mruby-bin-mrbc - mruby compiler executable
             mruby-test - mruby test
================================================

================================================
      Config Name: x86_64-pc-linux-gnu
 Output Directory: build/x86_64-pc-linux-gnu
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: i686-pc-linux-gnu
 Output Directory: build/i686-pc-linux-gnu
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: x86_64-apple-darwin14
 Output Directory: build/x86_64-apple-darwin14
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: i386-apple-darwin14
 Output Directory: build/i386-apple-darwin14
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: x86_64-w64-mingw32
 Output Directory: build/x86_64-w64-mingw32
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

================================================
      Config Name: i686-w64-mingw32
 Output Directory: build/i686-w64-mingw32
    Included Gems:
             mruby-io
             mruby-getopts - GNU getopt for mruby
             mruby-dir
             mruby-sprintf - standard Kernel#sprintf method
             mruby-time - standard Time class
             mruby-mtest
             mruby-cli - 0.0.4 - mruby cli utility
               - Binaries: mruby-cli
================================================

Writing releases/v0.0.4/mruby-cli-0.0.4-x86_64-pc-linux-gnu.tgz
Writing releases/v0.0.4/mruby-cli-0.0.4-i686-pc-linux-gnu.tgz
Writing releases/v0.0.4/mruby-cli-0.0.4-x86_64-apple-darwin14.tgz
Writing releases/v0.0.4/mruby-cli-0.0.4-i386-apple-darwin14.tgz
Writing releases/v0.0.4/mruby-cli-0.0.4-x86_64-w64-mingw32.tgz
Writing releases/v0.0.4/mruby-cli-0.0.4-i686-w64-mingw32.tgz
Writing releases/v0.0.4/mruby-cli-0.0.4.tgz
fpm is not installed. Please check your docker install.
```

drbrain commented:
I'm wondering why you want to import the mruby rake tasks atop your own.  I find [building mruby via subshell](https://github.com/fastly/mruby-optparse/blob/master/Rakefile#L12-L29) is less trouble and easier to use.  The same method could be used to run tasks inside docker.

zzak commented:
What is the point of running rake inside of a subshell.. in rake? :woah:

mruby knows how to compile mruby, and when we started building stuff with mruby, we leaned on the existing tasks to do the work for us. What you're doing with mruby-optparse is fine, but I don't think every gem doesn't want to maintain this much configuration.

The Rakefile is a generated file, that users shouldn't have to maintain, it's still there so the smaller the better.

But anyways, when I get a chance to finish [mruby-build](https://github.com/zzak/mruby-build), hopefully most of this stuff can be upstreamed and we don't have to bikeshed in each project. 

Back to this patch, the other thing is mruby-cli was generated.. with mruby-cli. So we have to update the setup.rb stuff so user apps can take advantage of these changes.

The downside to this patch is breaking up the tasks into separate files means having to generate them for the user too (on setup), more files means more things to diff when doing upgrades and such. So we have to consider that as well.

toch commented:
@drbrain Thanks for doing that. It's cleaner :+1:.

About package task, it seems that the last docker image is not the good one anymore (even if the Dockerfile in the repo is correct). @hone or @zzak could you give me access to the docker registry of mruby-cli image so I can fix it?

@drbrain as Zak explains, mruby-cli generates different files, including a Rakefile thanks to [`setup.rb`](https://github.com/hone/mruby-cli/blob/master/mrblib/mruby-cli/setup.rb). This is nearly the same than the Rakefile you improved (except a few things such as the `package` task). Could you also apply your changes to `setup.rb`? Thanks.

About, multiple files. @zzak I'm wondering if it wouldn't be the path to avoid to have twice the same code. If we can isolate the different task and make them generic, `setup.rb` can just copy the file, rather than generate it. What do you think?

drbrain commented:
I will update setup.rb soon.

@toch I must say I vastly prefer libraries over generated code.  It's a ton of work to update generated code, but library code updates are (within compatible versions) labor-free. Unfortunately I'm unsure how a library model can work with mruby-cli as a single binary.  (You _could_ do some bizarre stuff with `import` like I do in this PR, but it would likely quickly devolve into an unmaintainable mess.)

The reason I'm concerned about importing mruby's Rakefile is issues like [needing to work around mruby](https://github.com/drbrain/mruby-cli/blob/0472ccb01b9bc96bf65518d1643f78677f84d8e7/Rakefile#L14) or [having to tell users a task must run in docker](https://github.com/drbrain/mruby-cli/blob/0472ccb01b9bc96bf65518d1643f78677f84d8e7/Rakefile#L108-L113) (which [was broken](https://github.com/hone/mruby-cli/pull/43/commits/1683e022206babf45cdb9034c6a24496e8ff4a70)).

There's a few reasons I think separation will be better.

One is that it makes getting started as a rubyist familiar with rake much easier.  I can't `rake test` or `rake package`, instead I have to `docker-compose run test` but sometimes that doesn't work so I have to `docker-machine start default` then `eval $(docker-machine) env`.  I can probably make fully automated docker work with an imported mruby Rakefile, but it's a pain compared to having rake run docker for me.  (If you've never used docker mruby-cli has a bucket of mystery to drain before you can see "Hello World" show up on your terminal.)

Another reason I prefer a docker subshell is that I can't easily get tasks to run outside docker to set up state needed for tasks that must run inside docker.  For example, I want to generate some .rb files that end up in `mrblib/` without having to install a modern ruby inside Docker (thus having to manage my own image).  AFAICT this is impossible without running two commands currently because `docker-compose run` creates a dependency firewall.

With a docker subshell I could use regular rake dependencies to take care of that in a straightforward manner.  I could also set up tasks to easily upload my built CLI when docker was finished all from one rake task's dependencies.

For the above reasons I'm likely to switch my current mruby-cli build over to the docker subshell model because I will have less to teach my coworkers who will contribute to my CLI.  I can bring the instructions down to "download docker, install docker, run `rake`" from "download docker, install docker, start docker, load docker env, run `docker-compose run test`" … "oh, and if you reboot your machine you'll have to remember to start docker again and load the environment" (which they will forget because I forget).

The final one is that it's a separation of concerns.  Presently mruby-cli doesn't make any changes to mruby's Rakefile (besides the `local:ensure_in_docker` with this PR and similar code in existing mruby-cli).  This means I don't have any worries about upgrading to a new `mruby` with my existing mruby-cli.  Importing mruby's Rakefile makes it possible to accidentally couple to a given version of mruby meaning I'll have to update to a new mruby-cli to use a new mruby.  This would be annoying.

PS: Even rake has a `ruby` method for running ruby in a subshell.

drbrain commented:
Oh, the final thing I should add about importing mruby's Rakefile is that to do it in a rake-like fashion it involves features few people know about, like `import` and that it will invoke a task that depends on the imported file as this PR does to [make sure the download is unpacked](https://github.com/drbrain/mruby-cli/blob/0472ccb01b9bc96bf65518d1643f78677f84d8e7/Rakefile#L28-L29).

drbrain committed:
Update setup.rb to use updated Rakefile, etc

Also wrapped some lines at 80 columns to improve readability.

drbrain commented:
setup.rb updated

drbrain committed:
Remove too-late compiler check

This check's dependency is :all which runs the compilers so it can never
check what it needs to.

Unfortunately there's no place we can move this compiler check so that
it will work.  Having `task all: :compiler_check` won't work either
because the all task depends on the .o files which run the compiler
check.

I haven't found any tasks in mruby that we can hook into to run such
checks.  I looked extensively to find a place to hook-in for
cross-compiling third party libraries (like libcurl) and such a hook
doesn't exist.

In order to restore this functionality mruby-cli must not import mruby's
Rakefile.  Then we can have tasks that run before mruby is built (like
cross-compiling libraries or checking the cross-compiler environment).

drbrain commented:
See also @c6adb30 for a reason why mruby-cli should not import mruby's Rakefile.

toch commented:
@drbrain first of all, thanks for your patience and your answer.

Let me go through your answers first :).

### libraries vs generated code.

I agree with you. I think it would be cleaner to extract that part so we don't have to duplicate it. It's easy to forget to replicate it in the `setup.rb` and by doing so not providing it to the created mruby-cli app. As we would have the hand on them, there is no really a maintenance of a dependency, it's just a cleaner structure.

So far, my only current concern is as you said it well: "Unfortunately I'm unsure how a library model can work with mruby-cli as a single binary." In our case we need to consider: 1. `mruby-cli` the framework that allows us to scaffold the CLI app, and 2. the CLI itself. Considering that `mruby-cli` structure is nearly the same than the one of the CLI app it can scaffold, I don't think it would be a problem to opt for a library model. In a way we're already doing it as both depends on a few dependencies on mruby gems.

## refactoring of local:ensure_in_docker

Thanks a lot, it's cleaner and I learned something about Rake :).

## rake vs docker-compose

From a user point of view, I really like the idea. It would provide a better UX as we'd provide a very familiar set up, as you said.

## a docker subshell

I like what you suggest. In fact, it would allow us to use even more docker as a tool rather than an environment. There are options to do what you suggest from inside, but it's not simple and it would require in a certain case to run it with privileged mode what we don't want.

## about separation of concerns

agreed

I'm going to review your changes now.

toch commented on Rakefile:
```diff
@@ -84,239 +57,59 @@ desc "generate a release tarball"
-  # since we're in the mruby/
-  release_dir  = "releases/v#{APP_VERSION}"
-  release_path = Dir.pwd + "/../#{release_dir}"
+  release_path = File.expand_path "releases/v#{APP_VERSION}"
```
another good consequence :)
no more headache due to the surprising working directory

toch commented on Rakefile:
```diff
@@ -84,239 +57,59 @@ desc "generate a release tarball"
 task :release => :compile do
```
considering what you did with other long tasks, I think I'll extract that one too. For now, we can keep as it, there are already enough changed.

toch commented on mrblib/mruby-cli/setup.rb:
```diff
@@ -15,6 +15,10 @@ def run
 
+        create_dir_p("tasks")
+        write_file("tasks/package.rake", package_rake)
+        write_file("tasks/test.rake", test_rake)
```
I'm wondering if we shouldn't directly copy the task rather than generate them. It would avoid us to copy paste the code. I'm not a big fan of that. It's very brittle.

toch commented:
I've just commented a few lines. Not a lot to say except about the generation of code.

I still need to test the whole, I'll do that after my lunch.

toch commented:
There is an issue with the `clean` task. It seems that the `deep_clean` task is calling itself in a loop creating a loop bomb that froze my computer and made my X killed eventually.

For the task `package`, I got the following error:

```
NameError: undefined local variable or method `package_dir' for main:Object
/home/mruby/code/tasks/package.rake:125:in `block (3 levels) in <top (required)>'
/home/mruby/code/tasks/package.rake:122:in `each'
/home/mruby/code/tasks/package.rake:122:in `block (2 levels) in <top (required)>'
```

the variable [`package_dir`](https://github.com/hone/mruby-cli/pull/43/files#diff-298d604fa0852ec0a7a203b14e998eefR125) were [previously defined](https://github.com/hone/mruby-cli/pull/43/files#diff-52c976fc38ed2b4e3b1192f8a8e24cffL149) but it is no more in your changes.

The other tasks work as expected. The scaffold works also as expected except for the `clean` task.

Is it correct that for now, your changes don't implement your idea to run docker from rake?

drbrain commented:
I'll check the loop and missing `package_dir` soon.

I don't plan to implement running docker from rake here, it's too much change for one PR.

toch commented:
thanks

> I don't plan to implement running docker from rake here, it's too much change for one PR.

:+1: and I'd happy to help

drbrain committed:
Remove infinite loop in clean task

Previously we had a task rewriter that made sure certain tasks were run
in docker that prevented this infinite loop by accident.  After removing
the task rewriter the dependencies were circular (through #sh) and things
could run forever.  (Note that the `clean` task didn't change prior to
this commit.)

Rather than trying to unravel this infinite loop let's run `deep_clean`
directly via docker and let it do all the right things.  The `clean`
task remains for documentation purposes.  MRuby already defines it for
us, but now our users will know to hook into it.

drbrain commented:
Fixed the infinite loop by removing the looping.  Somehow the task rewriter caused this loop because I didn't change the clean task before @b6404e4.  See commit message for details.

drbrain committed:
Fix broken rename of `package_dir`

drbrain committed:
Update setup.rb with build system changes

drbrain commented:
This should be ready to go now, sorry for the delay.

toch commented:
Thanks @drbrain, could you rebase please?

I'm going to try the changes again to be sure.
